src/gurt/fault_inject.c

```diff
@@ -45,14 +45,66 @@
 
 #define FI_MAX_FAULT_ID 8192
 
+/* (1 << D_FA_TABLE_BITS) is the number of buckets of fa hash table */
+#define D_FA_TABLE_BITS		(13)
+
 #include <gurt/common.h>
+#include <gurt/hash.h>
+
+struct d_fault_attr {
+	d_list_t		fa_link;
+	struct d_fault_attr_t	fa_attr;
+};
+
+static struct d_fault_attr *
+fa_link2ptr(d_list_t *rlink)
+{
+	D_ASSERT(rlink != NULL);
+	return container_of(rlink, struct d_fault_attr, fa_link);
+}
+
+static bool
+fa_op_key_cmp(struct d_hash_table *htab, d_list_t *rlink, const void *key,
+	      unsigned int ksize)
+{
+	struct d_fault_attr *fa_ptr = fa_link2ptr(rlink);
+
+	D_ASSERT(ksize == sizeof(uint32_t));
+
+	return fa_ptr->fa_attr.fa_id == *(uint32_t *)key;
+}
+
+static void
+fa_op_rec_free(struct d_hash_table *htab, d_list_t *rlink)
+{
+	struct d_fault_attr	*ht_rec = fa_link2ptr(rlink);
+	int			 rc;
+
+	D_FREE(ht_rec->fa_attr.fa_argument);
+	rc = D_SPIN_DESTROY(&ht_rec->fa_attr.fa_lock);
+	if (rc != DER_SUCCESS)
+		D_ERROR("Can't destroy spinlock for fault id: %d\n",
+			ht_rec->fa_attr.fa_id);
+	D_FREE(ht_rec);
+}
+
+static bool
+fa_op_rec_decref(struct d_hash_table *htab, d_list_t *rlink)
```

Comment:
You still need a comment to explain how/why this code is misusing the hash table reference counting to be able to safely use decref without a corresponding addref function.